test/e2e/framework/metrics/latencies.go

```diff
@@ -0,0 +1,363 @@
+		if latency > latencyThreshold {
+			isBad = true
+			badMetrics++
+		}
+		if top > 0 || isBad {
+			top--
+			prefix := ""
+			if isBad {
+				prefix = "WARNING "
+			}
+			e2elog.Logf("%vTop latency metric: %+v", prefix, metrics.APICalls[i])
+		}
+	}
+	return badMetrics, metrics, nil
+}
+
+// VerifyLatencyWithinThreshold verifies whether 50, 90 and 99th percentiles of a latency metric are
+// within the expected threshold.
+func VerifyLatencyWithinThreshold(threshold, actual LatencyMetric, metricName string) error {
+	if actual.Perc50 > threshold.Perc50 {
+		return fmt.Errorf("too high %v latency 50th percentile: %v", metricName, actual.Perc50)
+	}
+	if actual.Perc90 > threshold.Perc90 {
+		return fmt.Errorf("too high %v latency 90th percentile: %v", metricName, actual.Perc90)
+	}
+	if actual.Perc99 > threshold.Perc99 {
+		return fmt.Errorf("too high %v latency 99th percentile: %v", metricName, actual.Perc99)
+	}
+	return nil
+}
+
+// ResetMetrics resets latency metrics in apiserver.
+func ResetMetrics(c clientset.Interface) error {
+	e2elog.Logf("Resetting latency metrics in apiserver...")
+	body, err := c.CoreV1().RESTClient().Delete().AbsPath("/metrics").DoRaw()
+	if err != nil {
+		return err
+	}
+	if string(body) != "metrics reset\n" {
+		return fmt.Errorf("Unexpected response: %q", string(body))
+	}
+	return nil
+}
+
+// Retrieves metrics information.
+func getMetrics(c clientset.Interface) (string, error) {
+	body, err := c.CoreV1().RESTClient().Get().AbsPath("/metrics").DoRaw()
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// Sends REST request to kube scheduler metrics
+func sendRestRequestToScheduler(c clientset.Interface, op, provider, cloudMasterName, masterHostname string) (string, error) {
+	opUpper := strings.ToUpper(op)
+	if opUpper != "GET" && opUpper != "DELETE" {
+		return "", fmt.Errorf("Unknown REST request")
+	}
+
+	nodes, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	// The following 4 lines are intended to replace framework.ExpectNoError(err)
+	if err != nil {
+		e2elog.Logf("Unexpected error occurred: %v", err)
+	}
+	gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
+
+	var masterRegistered = false
+	for _, node := range nodes.Items {
+		if system.IsMasterNode(node.Name) {
+			masterRegistered = true
+		}
+	}
+
+	var responseText string
+	if masterRegistered {
+		ctx, cancel := context.WithTimeout(context.Background(), SingleCallTimeout)
+		defer cancel()
+
+		body, err := c.CoreV1().RESTClient().Verb(opUpper).
+			Context(ctx).
+			Namespace(metav1.NamespaceSystem).
+			Resource("pods").
+			Name(fmt.Sprintf("kube-scheduler-%v:%v", cloudMasterName, ports.InsecureSchedulerPort)).
+			SubResource("proxy").
+			Suffix("metrics").
+			Do().Raw()
+
+		// The following 4 lines are intended to replace
+		// framework.ExpectNoError(err).
+		if err != nil {
+			e2elog.Logf("Unexpected error occurred: %v", err)
+		}
+		gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
```

Comment:
@andrewsykim replaced a framework.ExpectNoError(err) with the code for it here as well
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1360-L1372

Reply:
cc @timstclair

Reply:
Wrong Tim St Clair! cc @timothysc